feat: replace deprecated dependencies with their replacements (#5558)
renovatebot/renovate: docs/usage/configuration-options.md

}
```

### replacementName

Use this field to define the name of a replacement package.

> Should be used with `replacementVersion` (see example below).

Reviewer: ```suggestion
Must be used with `replacementVersion` (see example below).
```
See comment below on replacementVersion, seems both are required somehow.

New package rules can be added by editing [`replacements.ts`](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts)

### replacementVersion

Use this field to define the name of a replacement package.

> Should be used with `replacementVersion`.

Reviewer 2: Must it, really? Can't one just declare the replacementVersion without the replacementName?

> Should be used with `replacementVersion`.

Reviewer 3: I like having a fixed version. It's not like you could automatically take the latest version every time, e.g. if it's advanced to a new major release already

> Should be used with `replacementVersion`.

Reviewer: > Must it, really? Can't one just declare the replacementVersion without the replacementName?

IMO, it should be stated somewhere that both are required to build a replacement PR. If you have only one package rule with just replacementVersion, it will be a no-op.

As far as I understand, replacementName and replacementVersion could be extracted by a datasource (none of them in this PR) or set/overriden by package rule. It seems package rules with preset is the way to go.
If a datasource could extract one or the other, a package rule may override the extracted values. Say the version will be the case most often. There is no matchReplacementName, so if multiple replacement exist, you would override the version of an unpredictable replacementName.

Still, you're right in that package rules config are additive. One can define the replacementName and another one the replacementVersion. I don't know if this has happened with other params and what's the recommended way to document it.

> Should be used with `replacementVersion`.

Reviewer 3: I think it sounds safe for us to add a validation which makes sure they're in the same package rule

> Should be used with `replacementVersion`.

Author: Yeah, a change to a different dependency is most likely going to be a breaking change that will require manual intervention. Once the replacement PR is merged, a subsequent run of Renovate will upgrade to the latest version.

}
```

> ### replacementName

Reviewer 4: @JamieMagee The last two issues are not linked to be closed with this PR. I think this is because of a syntax error. I've checked the GitHub docs [^gh-docs], and I think this should work properly:

```markdown
Closes #2223, closes #5535, closes #3339
```

Relevant table from the GitHub Docs:

> Linked issue | Syntax | Example
> --------------- | ------ | ------
> Issue in the same repository | *KEYWORD* #*ISSUE-NUMBER* | `Closes #10`
> Issue in a different repository | *KEYWORD* *OWNER*/*REPOSITORY*#*ISSUE-NUMBER* | `Fixes octo-org/octo-repo#100`
> Multiple issues | Use full syntax for each issue | `Resolves #10, resolves #123, resolves octo-org/octo-repo#100`

[^gh-docs]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

Author: Thanks! I've update the body of the PR.

Use this field to define the name of a replacement package.
Must be used with `replacementVersion` (see example below).

> New package rules can be added by editing [`replacements.ts`](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts)

Reviewer 4: ```suggestion
You can suggest a new community package rule by editing [the `replacements.ts` file on the Renovate repository](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts) and opening a pull request.
```

- Clarify these are "community package rules".
- Use active voice for the "adding community rules to repository".
- Improve text for  the clickable link.
- End sentence with a full stop.

### replacementVersion

Use this field to define the name of a replacement package.
Must be used with `replacementVersion`.

> For example to replace the npm package `jade` with version `2.0.0` of the package `pug`:

Reviewer 4: ```suggestion
We recommend you also use the `matchDatasources` config option to specify the datasource, this prevents name collisions from other package managers.
For example to replace the npm package `jade` with version `2.0.0` of the package `pug`:
```

Say you have Python and npm package files in your repo, and you have a package called `pug` that exists on both package manager, I think you'd run into conflicts then?

Do we think the chance of collisions is small enough that we don't need to warn about it?

Reviewer 3: I think that the current docs are adequate

The field supports multiple URLs however it is datasource-dependent on whether only the first is used or multiple.

## replacementName

## replacementVersion

> 

Reviewer 3: Needs expanding, including if both are mandatory or only `replacementName`.

Reviewer 2: Fair point. Other changes indicate that the both of them are required. Allowing the both of them to be optional would allow for easier integration with existing available metadata, like [Maven's relocation](https://maven.apache.org/guides/mini/guide-relocation.html) information. With that, the `groupId`, `artifactId` and `version` can all be `null`.

Reviewer 2: Would it make sense to add these fields to the `validMatchFields` of [the regex manager](lib/manager/regex/index.ts)? I already have scenarios where, based on certain regex matches/capture groups, I wanna compose a replacement version string.

Author: @rarkins Can do

@astellingwerf Yes, but I would like to get some form of this merged in a _very_ limited state, and then build on top of that. Adding support for the regex manager should come in a later iteration, in my opinion. This is already a large, long-lived, pull request—though that is, admittedly, my fault.

Reviewer 2: Would you then recommend me to hold off on my MR, and create a separate issue and MR to add replacement support for Gradle/Maven/Java/...?

Author: @astellingwerf Sorry, I missed your response earlier. I agree with holding off for now, and creating followup issues/PRs
